chips/apollo3/src/flash_ctrl.rs

```diff
@@ -0,0 +1,276 @@
+    Read,
+    Write,
+    Erase,
+}
+
+pub struct FlashCtrl<'a> {
+    flash_client: OptionalCell<&'a dyn hil::flash::Client<FlashCtrl<'a>>>,
+    read_buf: TakeCell<'static, Apollo3Page>,
+    write_buf: TakeCell<'static, Apollo3Page>,
+
+    deferred_call: DeferredCall,
+    op: Cell<Operation>,
+}
+
+impl<'a> FlashCtrl<'a> {
+    pub fn new() -> Self {
+        FlashCtrl {
+            flash_client: OptionalCell::empty(),
+            read_buf: TakeCell::empty(),
+            write_buf: TakeCell::empty(),
+            deferred_call: DeferredCall::new(),
+            op: Cell::new(Operation::None),
+        }
+    }
+}
+
+impl<C: hil::flash::Client<Self>> hil::flash::HasClient<'static, C> for FlashCtrl<'_> {
+    fn set_client(&self, client: &'static C) {
+        self.flash_client.set(client);
+    }
+}
+
+impl hil::flash::Flash for FlashCtrl<'_> {
+    type Page = Apollo3Page;
+
+    fn read_page(
+        &self,
+        page_number: usize,
+        buf: &'static mut Self::Page,
+    ) -> Result<(), (ErrorCode, &'static mut Self::Page)> {
+        if page_number >= FLASH_MAX_PAGES {
+            return Err((ErrorCode::INVAL, buf));
+        }
+
+        if self.op.get() != Operation::None {
+            return Err((ErrorCode::BUSY, buf));
+        }
+
+        if self.deferred_call.is_pending() {
+            return Err((ErrorCode::BUSY, buf));
+        }
+
+        let addr = (page_number * PAGE_SIZE) as u32;
+        let addr_ptr = addr as *mut u32;
+
+        // Call `uint32_t flash_util_read_word(uint32_t *)` in the ROM code.
+        let ptr = 0x08000075 as *const ();
+        let f: extern "C" fn(*mut u32) -> u32 = unsafe { core::mem::transmute(ptr) };
+
+        for i in 0..(PAGE_SIZE / 4) {
+            let val = (f)(addr_ptr.wrapping_add(i)).to_le_bytes();
+            let offset = i * 4;
+
+            buf[offset] = val[0];
+            buf[offset + 1] = val[1];
+            buf[offset + 2] = val[2];
+            buf[offset + 3] = val[3];
+        }
+
+        self.read_buf.replace(buf);
+        self.op.set(Operation::Read);
+        self.deferred_call.set();
+        Ok(())
+    }
+
+    fn write_page(
+        &self,
+        page_number: usize,
+        buf: &'static mut Self::Page,
+    ) -> Result<(), (ErrorCode, &'static mut Self::Page)> {
+        if page_number >= FLASH_MAX_PAGES {
+            return Err((ErrorCode::INVAL, buf));
+        }
+
+        if self.op.get() != Operation::None {
+            return Err((ErrorCode::BUSY, buf));
+        }
+
+        if self.deferred_call.is_pending() {
+            return Err((ErrorCode::BUSY, buf));
+        }
+
+        let addr = (page_number * PAGE_SIZE) as u32;
+        let addr_ptr = addr as *mut u32;
+
+        // Call `int flash_program_main(uint32_t, uint32_t *, uint32_t *, uint32_t)` in the ROM code.
+        let ptr = 0x08000055 as *const ();
+        let f: extern "C" fn(u32, *mut u32, *mut u32, u32) -> c_int =
+            unsafe { core::mem::transmute(ptr) };
```

Comment:
I have not conducted an in-depth review yet, but this is a type of unsafe code that we don't yet have in Tock. Regardless of a deeper discussion about calling into ROM code, this should definitely be a cast to an unsafe function:

```suggestion
        // Call `int flash_program_main(uint32_t, uint32_t *, uint32_t *, uint32_t)` in the ROM code.
        let rom_flash_program_main: extern "C" fn(u32, *mut u32, *mut u32, u32) -> c_int =
            unsafe { core::mem::transmute(0x08000055 as *const ()) };
```

We'll also want to avoid having variable names like `ptr` and `f` in scope, _especially_ in such safety-critical code.